test/firmata.test.js

```diff
@@ -2050,6 +2050,131 @@ describe("Board: instances", function() {
     });
 
   });
+
+  describe("parser", function() {
+
+    beforeEach(function() {
+      board.currentBuffer = [];
+    });
+
+    it("must parse a command from the beginning of a data packet", function (done) {
+      var spy = sandbox.spy();
+      var incoming = [REPORT_VERSION, 0x02, 0x03];
+      board.versionReceived = false;
+      board.on("reportversion", spy);
+      transport.emit("data", incoming);
+      assert.equal(spy.callCount, 1);
+      done();
+    });
+
+    it("must parse a command from the middle of a data packet", function (done) {
+      var spy = sandbox.spy();
+      // includes: analog input, report version, query firmware (incomplete)
+      var incoming = [
+        0xe0, 0x07, 0x07, 0xf9, 0x02, 0x05, 0xf0, 0x79, 0x02, 0x05, 0x53, 0x00, 0x74, 0x00, 0x61,
+        0x00, 0x6e, 0x00, 0x64, 0x00
+      ];
+      board.versionReceived = false;
+      board.on("reportversion", spy);
+      transport.emit("data", incoming);
+      assert.equal(spy.callCount, 1);
+      done();
+    });
+
+    it("must not emit command events until REPORT_VERSION is received", function (done) {
+      var spyAnalog = sandbox.spy();
+      var spyVersion = sandbox.spy();
+      // includes: analog input, report version, query firmware (incomplete) and junk
+      // between analog input and report version
+      var incoming = [
+        0xe0, 0x00, 0x71, 0xf9, 0x02, 0x05, 0xf0, 0x79, 0x02, 0x05, 0x53, 0x00, 0x74,
+        0x00, 0x61, 0x00, 0x6e, 0x00, 0x64, 0x00
+      ];
+      board.versionReceived = false;
+      board.on("analog-read-0", spyAnalog);
+      board.on("reportversion", spyVersion);
+      transport.emit("data", incoming);
+      assert.equal(spyAnalog.callCount, 0);
+      assert.equal(spyVersion.callCount, 1);
+      done();
+    });
+
+    it("must parse multiple commands from a single packet", function (done) {
+      var spyAnalog = sandbox.spy();
+      var spyVersion = sandbox.spy();
+      // includes: report version, analog input, query firmware (incomplete) and junk
+      // between analog input and report version
+      var incoming = [
+        0xf9, 0x02, 0x05, 0xe0, 0x00, 0x71, 0xf0, 0x79, 0x02, 0x05, 0x53, 0x00, 0x74,
+        0x00, 0x61, 0x00, 0x6e, 0x00, 0x64, 0x00
+      ];
+      board.versionReceived = false;
+      board.on("reportversion", spyVersion);
+      board.on("analog-read-0", spyAnalog);
+      transport.emit("data", incoming);
+      assert.equal(spyVersion.callCount, 1);
+      assert.equal(spyAnalog.callCount, 1);
+      done();
+    });
+
+    it("must parse a complete sysex command after an incomplete sysex command", function (done) {
+      var spy = sandbox.spy();
+      // includes: query firmware (incomplete sysex), pin state response (pin 2)
+      var incoming = [
+        0xf0, 0x79, 0x02, 0x05, 0x53, 0x00, 0x74, 0x00, 0x61, 0x00, 0x6e, 0x00, 0x64, 0x00,
+        0xf0, 0x6e, 0x02, 0x01, 0x01, 0xf7
+      ];
+      board.versionReceived = true;
```

Comment:
any of the following tests where this is set to `true` will break if `REPORT_VERSION` is not sent.